src/jdk.compiler/share/classes/com/sun/tools/javac/comp/LambdaToMethod.java

```diff
@@ -916,809 +952,335 @@ private JCExpression makeIndyCall(DiagnosticPosition pos, Type site, Name bsmNam
+            StringBuilder buf = new StringBuilder();
+            buf.append(names.lambda);
+            // Append the name of the method enclosing the lambda.
+            buf.append(syntheticMethodNameComponent(owner));
+            buf.append('$');
+            // Append a hash of the disambiguating string : enclosing method
+            // signature, etc.
+            String disam = serializedLambdaDisambiguation(owner);
+            buf.append(Integer.toHexString(disam.hashCode()));
+            buf.append('$');
+            // The above appended name components may not be unique, append
+            // a count based on the above name components.
+            buf.append(kInfo.syntheticNameIndex(buf, 1));
+            String result = buf.toString();
+            //System.err.printf("serializedLambdaName: %s -- %s\n", result, disam);
+            return names.fromString(result);
         }
 
         /**
-         * This class is used to store important information regarding translation of
-         * lambda expression/method references (see subclasses).
+         * Translate a symbol of a given kind into something suitable for the
+         * synthetic lambda body
          */
-        abstract class TranslationContext<T extends JCFunctionalExpression> {
-
-            /** the underlying (untranslated) tree */
-            final T tree;
-
-            /** points to the adjusted enclosing scope in which this lambda/mref expression occurs */
-            final Symbol owner;
-
-            /** the depth of this lambda expression in the frame stack */
-            final int depth;
-
-            /** the enclosing translation context (set for nested lambdas/mref) */
-            final TranslationContext<?> prev;
-
-            /** list of methods to be bridged by the meta-factory */
-            final List<Symbol> bridges;
-
-            TranslationContext(T tree) {
-                this.tree = tree;
-                this.owner = owner(true);
-                this.depth = frameStack.size() - 1;
-                this.prev = context();
-                ClassSymbol csym =
-                        types.makeFunctionalInterfaceClass(attrEnv, names.empty, tree.target, ABSTRACT | INTERFACE);
-                this.bridges = types.functionalInterfaceBridges(csym);
+        VarSymbol translate(final VarSymbol sym, LambdaSymbolKind skind) {
+            VarSymbol ret;
+            boolean propagateAnnos = true;
+            switch (skind) {
+                case CAPTURED_VAR:
+                    Name name = (sym.flags() & LOCAL_CAPTURE_FIELD) != 0 ?
+                            sym.baseSymbol().name : sym.name;
+                    ret = new VarSymbol(SYNTHETIC | FINAL | PARAMETER, name, types.erasure(sym.type), translatedSym);
+                    propagateAnnos = false;
+                    break;
+                case LOCAL_VAR:
+                    ret = new VarSymbol(sym.flags() & FINAL, sym.name, sym.type, translatedSym);
```

Comment:
shouldn't this symbol be synthetic too? same for the param below

Reply:
I suppose local variables and parameters come from "real" symbols in the source code. We could probably do either way, but for now I preferred not to touch this stuff